GPII 3340 disable basic auth and client certs

stepanstipl opened the pull request:
This PR is a pre-req to GPII-3340. It:
- Updates Helm to 0.6.0 (needed for diabling tiller install)
- Disables tiller install by helm_release resource, this prevents nasty cluster modification even in plan mode (see https://github.com/mcuadros/terraform-provider-helm/issues/121 for details)
- Adds support to set/diable basic auth
- Adds support to disable issuing client certs
- Removes create_before_destroy (this can't work for GKE cluster)
- Minor terraform fmt changes
- Fix to reinstall tiller if the cluster is recreated
- Add 30s timout for uninstalling tiller (now parametrised)

~**WIP** This is not complete yet as I'm still working on the issue of correctly redeploying tiller when the cluster is recreated (because of change like disabling client certs)., still want to do some more testing.~
Refer to https://github.com/gpii-ops/gpii-infra/pull/154 for details about tested scenarios.

mrtyler commented on modules/gke-cluster/variables.tf:
```diff
@@ -70,10 +70,25 @@ variable "logging_service" {
+}
+
+variable "master_auth_password" {
+  default     = ""
```
What happens if `master_auth_(username|password)` is the default value, i.e. empty string? Might be worth a comment (perhaps in the 'description' field).

stepanstipl replied:
Yeah good point, I'll add a comment - when empty, it actually disables the basic auth (that's what we want).

stepanstipl commented:
Added a comment addressing Tyler's remark and code to get tiller reinstalled.

This is not super clean, as it'll try to install tiller twice - 1st time, and once more after that, but I wasn't really able to come up with anything better given it's implemented via null_resource & provisioners.

Alternative I came up with would be to use cluster's `createTime` (from `gcloud containter clusters describe`) as a trigger, but while that does not have the problem with 2 installs, it has other disadvantages:
- having to introduce and pass 2 new parameters to helm-initializer - cluster name and zone
- will not work when someone removes or breakes tiller without recreating the cluster (the implemented approach will)

Happy to hear suggestions how to improve this. And I want to do some more testing before merging (to cover more scenarios).

amatas commented:
LGTM so far, I can not think on any better alternative to avoid multiple installations of the Tiller. Anyway this seems a good start.

mrtyler commented on modules/gke-cluster/variables.tf:
```diff
@@ -74,6 +74,8 @@ variable "dashboard_disabled" {
   description = "Disable Kubernetes Dashboard"
 }
 
+# Setting both password and username to empty strings (module default)
```
I think this will be easier for a user to notice if it's in the Description.

stepanstipl replied:
How would you prefer this? In username? Password? Both?

The behavior here mirrors the one of TF's `container_cluster` and the one of GCP api & `gcloud` tool, not sure we should do much more explanation here. See the "basic auth" section at https://cloud.google.com/sdk/gcloud/reference/container/clusters/create - the behavior is quite complex.

mrtyler replied:
I was thinking that you keep what you have as a comment, but in the description for each variable.

I agree that it's not our job to exhaustively document the gcloud API here :). I think a hint is warranted since I can maybe guess the default behavior but it is not trivially clear to me. Contrast this case with `zone` (pretty clear what it's for; no comment needed) or `addons` (pretty clear that I need to look at the API docs if I want to know more).

Move it to description, leave it as-is, or remove it -- I leave it up to you.

stepanstipl replied:
thanks for the clarification, I'll move it to descriptions of both username & password

mrtyler commented on modules/helm-initializer/main.tf:
```diff
@@ -26,7 +26,18 @@ data "template_file" "tiller_rbac" {
+  program = [
+    "bash",
+    "-c",
+    "REPLICAS=$$(kubectl get deploy tiller-deploy -n ${var.tiller_namespace} -o jsonpath='{.status.readyReplicas}'); jq -n --arg replicas \"$$REPLICAS\" '{readyReplicas:$$replicas}'"]
```
Should these commands be joined with `&&` instead of `;`? What happens to the trigger if the kubectl command fails, e.g. due to something dumb like a connection timeout?

mrtyler commented on modules/helm-initializer/main.tf:
```diff
@@ -26,7 +26,18 @@ data "template_file" "tiller_rbac" {
+  program = [
+    "bash",
+    "-c",
+    "REPLICAS=$$(kubectl get deploy tiller-deploy -n ${var.tiller_namespace} -o jsonpath='{.status.readyReplicas}'); jq -n --arg replicas \"$$REPLICAS\" '{readyReplicas:$$replicas}'"]
```
What advantage do we get from using `jq` here and returning json vs returning readyReplicas as a bare string? AFAIK trigger doesn't need JSON as input.

mrtyler replied:
Ok, thanks.

mrtyler commented:
LGTM

mrtyler commented:
> it'll try to install tiller twice - 1st time, and once more after that

I don't think I understand why this happens, so I don't have suggestions on approach. Might be worth a brief parking lot discussion in case the team can figure something out?

I agree that we want to minimize extra work (like re-checking whether tiller is installed), but your alternative doesn't seem better (favor correctness over speed). I'm ok with the current version for now.

natarajaya commented on modules/helm-initializer/main.tf:
```diff
@@ -54,7 +65,8 @@ helm reset --force \
 --tls-cert=${local_file.helm_cert.filename} \
---tls-key=${local_file.helm_key.filename}
+--tls-key=${local_file.helm_key.filename} \
+--tiller-connection-timeout 30
```
Since you're changing default value for this, should not it be configurable?

stepanstipl replied:
good point, thanks, I'll make it parametrised

natarajaya commented:
LGTM, just a few minor issues.

> I don't think I understand why this happens, so I don't have suggestions on approach. Might be worth a brief parking lot discussion in case the team can figure something out?

+1

stepanstipl commented on modules/helm-initializer/main.tf:
```diff
@@ -26,7 +26,18 @@ data "template_file" "tiller_rbac" {
+
 resource "null_resource" "install_tiller" {
+  triggers {
+    tiller_ready_replicas = "${data.external.tiller_status.result.readyReplicas}"
```
Thanks Sergey, the tiller deploy is defined as follows: 
```
spec:
...
  replicas: 1
...
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
...
```
so it should not happen normally that there would be more than 1 replica. You're right it can happen in case of an update, but I assume that it should be settled back to 1 on the next TF run. 

I preffered simplicity of implementation in this case, as what would happen if the replicase would be `2` for example, the tiller would just be recreated, which is undesirable, but not really a problem (and as mentioned happens anyway once on the initial cluster creation). Otherwise it can get more cmplicated as there are scenarios when kubectl can't connect or tiller-deply does not exist at all, or it might not have the `readyReplicas` field yet - and after thinking about this I leaned towards keeping it simple.

natarajaya replied:
Alright, thanks for the explanation!

stepanstipl commented:
@mrtyler 
> I don't think I understand why this happens, so I don't have suggestions on approach. Might be worth a brief parking lot discussion in case the team can figure something out?
> 
> I agree that we want to minimize extra work (like re-checking whether tiller is installed), but your alternative doesn't seem better (favor correctness over speed). I'm ok with the current version for now.

This happens as the trigger mechanism is not based on the state, but rather on the change of the state. So:
- **1st run**: resource does not exist, the trigger is evaluated to `""` (empty) -> resource is created
- **2nd run**: resource does exist, but the trigger changed to `"1"` -> resource is recreated
- **3rd and subsequent runs**: resource does exist, trigger stays `"1"` -> no changes
... unless the tiller is deleted or stops working, then the trigger is changed and will cause resource recreation and the behavior described above will repeat once more

@mrtyler @natarajaya I tried to explain the behaviour here, please add an item to standup if you want to discuss further - I'm happy to.

stepanstipl commented:
I believe I've addressed all the comments above as well as added a comment describing the trigger behavior that we discussed yesterday at standup.

While I have run into various issues during testing, none I believe were related to this change. For details about scenarios I've tested please refer to https://github.com/gpii-ops/gpii-infra/pull/154.

I'll also update this PR with a reference to the upstream PR once that is opened.

stepanstipl commented:
PRs against upstream have been opened:
https://github.com/exekube/exekube/pull/107 - Update helm provider to 0.6.0 and Disable tiller deployment by Helm provider
https://github.com/exekube/exekube/pull/108 - Fix tiller redeployment after cluster is recreated
https://github.com/exekube/exekube/pull/109 - Add support for setting basic auth and disabling client certificates
